lib/moab/storage_object_validator.rb

```diff
@@ -0,0 +1,196 @@
+      # if we only have sequential version directories uder the root object path, proceed
+      # to check for expected version subdirs, and to check contents of the data dir
+
+      results.concat correctly_formed_moab if results.empty?
+
+      results
+    end
+    # TODO: Figure out which methods should be public
+
+    private
+
+    def check_for_only_sequential_version_dirs
+      results = []
+
+      sub_dirs(storage_obj_path).each_with_index do |dir_name, index|
+        expected_vers_num = index + 1 # version numbering starts at 1, array indexing starts at 0
+        begin
+          if dir_name[1..-1].to_i != expected_vers_num
+            results << result_hash(VERSIONS_NOT_IN_ORDER, sub_dirs(storage_obj_path))
+            break
+          end
+        rescue ArgumentError
+          results << result_hash(VERSIONS_NOT_IN_ORDER, sub_dirs(storage_obj_path))
+        end
+      end
+
+      unless files_in_dir(storage_obj_path).empty?
+        results << result_hash(FILES_IN_VERSION_DIR, files_in_dir(storage_obj_path))
+      end
+
+      results
+    end
+
+    def correctly_formed_moab
+      results = []
+
+      version_directories = sub_dirs(storage_obj_path)
+      version_directories.each do |version_dir|
+        version_path = "#{storage_obj_path}/#{version_dir}"
+        version_sub_dirs = sub_dirs(version_path)
+        # Need to add before_result_size and after_result_size because if everything is expected
+        # and good, then no error message will be added to the result array, so if it stays the same
+        # size then it is ready to be checked for the data_sub_dirs and manfist_xml
+        before_result_size = results.size
+        results.concat check_sub_dirs(version_sub_dirs, version_dir, EXPECTED_VERSION_SUB_DIRS)
+        after_result_size = results.size
+        # don't bother checking the data sub dir if we didn't find the expected two subdirs
+
+        # Before it was result.size === 0, however for each version that is not correctly constructed,
+        # error messages will be concat to the array. By the time the loop reaches a version that is
+        # correctly constructe the size will stay the same as before (because it has the old error messages)
+        # so we must make sure before and after result size are the same - aka no error message was added
+        # and the version directory was correctly constructed
+
+        if before_result_size == after_result_size
+          data_dir_path = "#{version_path}/#{version_sub_dirs[0]}"
+          data_sub_dirs = sub_dirs(data_dir_path)
+          results.concat check_sub_dirs(data_sub_dirs, version_dir, EXPECTED_DATA_SUB_DIRS)
+          results.concat check_manifest_files(version_path, version_dir)
+        end
+      end
+
+      results
+    end
+
+    def check_sub_dirs(sub_dirs, version, required_sub_dirs)
+      results = []
+      sub_dir_count = sub_dirs.size
+      if sub_dir_count == 2
```

Comment:
i think here you want to substitute:  `required_sub_dirs.size` for hardcoded constant `2` -- because in this context, it could be any number, no??